mcs/build/gensources.cs

```diff
@@ -0,0 +1,411 @@
+using System;
+using System.Text;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+public static class Program {
+    public static int Main (string[] _args) {
+        var args = new List<string> (_args);
+        bool useStdout = false, showHelp = false, strictMode = false;
+
+        for (int i = 0; i < args.Count; i++) {
+            var arg = args[i];
+            if (!arg.StartsWith ("-"))
+                continue;
+
+            switch (arg) {
+                case "-?":
+                case "--help":
+                case "-h":
+                    showHelp = true;
+                    break;
+                case "--trace":
+                case "--trace1":
+                    SourcesParser.TraceLevel = 1;
+                    break;
+                case "--trace2":
+                    SourcesParser.TraceLevel = 2;
+                    break;
+                case "--trace3":
+                    SourcesParser.TraceLevel = 3;
+                    break;
+                case "--trace4":
+                    SourcesParser.TraceLevel = 4;
+                    break;
```

Comment:
not sure we really need four logging levels for this :) I'd consolidate it into a single `--verbose`